### PR TITLE
WIP: Converts to using bundle over dereference from json-schema-ref-parser

### PR DESCRIPTION
**Pull request by scalvert**
The `bundle` method allows for inlining of refs within the supplied schema. This helps resolve schemas with circular references, as those references now become inlined and thus, non-circular. 

⚠️ Note: this change results in a change to the public API, and therefore would require a major version bump.

Resolves: #644

### CODE REVIEW

**Comment by scalvert** [comment]
@pateketrueke I'm not super familiar with the repo, but think inlining the refs would be a nice addition since it eliminates the circular reference problem. What do you think? Should we try to move this forward by identifying the issue in the 2 tests and fixing?

File: src/lib/index.js [hunk 1]
```diff
@@ -108,16 +108,7 @@ jsf.generateYAML = (schema, refs) => renderYAML(
...
 
-jsf.resolveWithContext = (schema, refs, cwd) => {
-  if (typeof refs === 'string') {
-    cwd = refs;
-    refs = {};
-  }
-
-  // normalize basedir (browser aware)
-  cwd = cwd || (typeof process !== 'undefined' ? process.cwd() : '');
-  cwd = `${cwd.replace(/\/+$/, '')}/`;
```
[the remark this answers is not in the document]
Comment (scalvert) [reply src/lib/index.js:119]: Good question. Do we have tests that cover this case?

[the remark this answers is not in the document]
Comment (pateketrueke) [reply src/lib/index.js:119]: seems like we don't have a test for those cases, sorry

**Comment by pateketrueke** [comment]
I've been playing a bit with these changes, the affected tests are:

- tests/schema/core/issues/issue-529.json
- tests/schema/core/issues/issue-446.json

They failed on their first attemp to validate the schema against the sampled output.

I still don't get why they're failing since they have only local definitions, it looks like using `bundle` effectively drops all references by inlining them, and that kinda brokes how the references are resolved during sample generation of `anyOf|oneOf` keywords.

Let me see if I can fix how those keywords would behave, if that works I think we could stick then to `bundle` and probably remove all the code that resolve `$ref` schemas.

**Comment by pateketrueke** [comment]
I also found that this code just works fine:

```js
import $RefParser from 'json-schema-ref-parser';

const schema = {
  "$schema": "http://json-schema.org/draft-04/schema#",
  "title": "Static Analysis Results Format (SARIF) Version 2.1.0-rtm.5 fragment",
  "id": "https://sarif/result-fragment.json",
  "description": "A dynamic fragment of a SARIF schema.",
  "allOf": [
    {
      "$ref": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json#/definitions/result"
    }
  ]
};

$RefParser.dereference(schema).then(console.log);
```

However it does not play well with jsf, so I'm now more confident that we should move to `bundle` instead.

I'll be working on patching the offended keywords, send me good vibes! :beers:

**Comment by pateketrueke** [comment]
For some reason, with `bundle` there are leftovers of `anyOf|oneOf|allOf` props during traversing, they weren't resolved properly and I cannot determine the root-cause yet...

However, I find a workaround:

```diff
diff --git a/src/lib/core/traverse.js b/src/lib/core/traverse.js
index a0c21a9..a0fc160 100644
--- a/src/lib/core/traverse.js
+++ b/src/lib/core/traverse.js
@@ -18,6 +18,10 @@ function getMeta({ $comment: comment, title, description }) {
 function traverse(schema, path, resolve, rootSchema) {
   schema = resolve(schema, null, path);

+  if (schema && (schema.oneOf || schema.anyOf || schema.allOf)) {
+    schema = resolve(schema, null, path);
+  }
+
   if (!schema) {
     return;
   }
diff --git a/src/lib/index.js b/src/lib/index.js
index 2d50056..ed90684 100644
--- a/src/lib/index.js
+++ b/src/lib/index.js
@@ -138,7 +138,7 @@ jsf.resolveWithContext = (schema, refs, cwd) => {
   };

   return $RefParser
-    .dereference(cwd, schema, {
+    .bundle(cwd, schema, {
       resolve: {
         file: { order: 100 },
         http: { order: 200 },
```

WIth these exact changes all tests are passing now.

Please revert the code you deleted.

Thank you!

**Pull request closed without merging**